src/main/java/pl/futurecollars/invoicing/service/InvoiceService.java

```diff
@@ -0,0 +1,77 @@
+package pl.futurecollars.invoicing.service;
+
+import java.math.BigDecimal;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import lombok.Data;
+import pl.futurecollars.invoicing.db.InMemoryDatabase;
+import pl.futurecollars.invoicing.model.Invoice;
+import pl.futurecollars.invoicing.model.InvoiceEntry;
+
+@Data
+public class InvoiceService {
+
+    private final InMemoryDatabase database;
+
+    public BigDecimal getTotalNet(Invoice invoice) {
+        return invoice.getEntries()
+            .stream()
+            .map(InvoiceEntry::getPrice)
+            .reduce(BigDecimal.valueOf(0), BigDecimal::add);
+    }
+
+    public BigDecimal getTotalTaxValue(Invoice invoice) {
+        return invoice.getEntries()
+            .stream()
+            .map(InvoiceEntry::getVatValue)
+            .reduce(BigDecimal.valueOf(0), BigDecimal::add);
+    }
+
+    public BigDecimal getTotalGross(Invoice invoice) {
+        return getTotalNet(invoice).add(getTotalTaxValue(invoice));
+    }
+
+    public List<Invoice> filterByIssuerName(String issuerName) {
+        return database.getAll()
+            .stream()
+            .filter((Invoice invoice) -> invoice.getIssuer().getName().equals(issuerName))
+            .collect(Collectors.toList());
+    }
+
+    public List<Invoice> filterByReceiverName(String receiverName) {
+        return database.getAll()
+            .stream()
+            .filter((Invoice invoice) -> invoice.getReceiver().getName().equals(receiverName))
```

Comment:
better is to switch comparison of the `String` objects to avoid `NullPointerException`

`(Invoice invoice) -> receiverName.equals(invoice.getReceiver().getName())`

because it is less possible to have `receiverName` set to `null`

Reply:
Ok I see. Please correct me if I understand it wrong: e.g. `string1.equals(string2)` if `string1` is null then we will get `NullPointerException`, if `string2` is null we will get `false`?

Reply:
Corrected